pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -580,30 +580,56 @@ public synchronized void redeliverUnacknowledgedMessages(Consumer consumer, List
     @Override
     public void addUnAckedMessages(int numberOfMessages) {
         // don't block dispatching if maxUnackedMessages = 0
-        if(maxUnackedMessages <= 0) {
+        if (maxUnackedMessages <= 0) {
             return;
         }
         int unAckedMessages = TOTAL_UNACKED_MESSAGES_UPDATER.addAndGet(this, numberOfMessages);
         if (unAckedMessages >= maxUnackedMessages
                 && BLOCKED_DISPATCHER_ON_UNACKMSG_UPDATER.compareAndSet(this, FALSE, TRUE)) {
+            // block dispatcher if it reaches maxUnAckMsg limit
             log.info("[{}] Dispatcher is blocked due to unackMessages {} reached to max {}", name,
                     TOTAL_UNACKED_MESSAGES_UPDATER.get(this), maxUnackedMessages);
+        } else if (topic.getBrokerService().isBrokerDispatchingBlocked()
+                && BLOCKED_DISPATCHER_ON_UNACKMSG_UPDATER.get(this) == TRUE) {
+            // unblock dispatcher: if dispatcher is blocked due to broker-unackMsg limit and if it ack back enough
+            // messages
+            if (TOTAL_UNACKED_MESSAGES_UPDATER.get(this) < (topic.getBrokerService().maxUnackedMsgsPerDispatcher / 2)) {
+                if (BLOCKED_DISPATCHER_ON_UNACKMSG_UPDATER.compareAndSet(this, TRUE, FALSE)) {
+                    // it removes dispatcher from blocked list and unblocks dispatcher by scheduling read
+                    topic.getBrokerService().unblockDispatchersOnUnAckMessages(Lists.newArrayList(this));
+                }
+            }
         } else if (BLOCKED_DISPATCHER_ON_UNACKMSG_UPDATER.get(this) == TRUE
                 && unAckedMessages < maxUnackedMessages / 2) {
+            // unblock dispatcher if it acks back enough messages
             if (BLOCKED_DISPATCHER_ON_UNACKMSG_UPDATER.compareAndSet(this, TRUE, FALSE)) {
                 log.info("[{}] Dispatcher is unblocked", name);
                 topic.getBrokerService().executor().submit(() -> readMoreEntries());
```

Comment:
dispatcher can be part of `blockedDispatchersList` only if brokerService is blocked (`topic.getBrokerService().isBrokerDispatchingBlocked()`) which we are checking first in ln#592. if brokerService is not blocked it means `blockedDispatchersList` is empty but this dispatcher was blocked because it reached individual max limit and should be unblock after received enough acked messages.